README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bandsfdp

<!-- badges: start -->
<!-- badges: end -->

This package provides functions that compute upper prediction bounds on
the FDP in competition-based setups (see [Ebadi et
al. (2022)](https://arxiv.org/abs/2302.11837)). Such setups include
target-decoy competition (TDC) in computational mass spectrometry and
the knockoff construction in regression. Note we typically use the
terminology of TDC throughout.

In (single-decoy) TDC, each hypothesis is associated to a winning score
and a label
($1$
for a target win and
$-1$
for a decoy win). Functions in this package assume that the hypotheses
are ordered in decreasing order of winning scores (with ties broken at
random).

The functions `tdc_sb()` and `tdc_ub()` give an upper prediction bound
on the FDP in TDC’s discovery list. Given TDC’s rejection threshold, the
target/decoy labels, and a desired confidence level
$1 - \gamma$,
these functions return a real number
$[\eta$
such that the FDP in the list of discoveries is
$\leq \eta$
with probability
$\geq 1 - \gamma$.

The function `sim_bound()` provides simultaneous bounds on the FDP. It
computes an upper prediction bound on the FDP of target wins among the
top
$k$
hypotheses of TDC (the hypotheses of the
$k$
largest winning scores), for each
$k = 1,\ldots,n$
where
$n$
is the total number of hypotheses. Similarly, the function `gen_bound()`
provides a bound on the FDP among target wins in an arbitrary set
$R$
of hypotheses of TDC.

Note that upper prediction bounds are derived from upper prediction
bands. In particular, the bounds in this package are derived from the
standardized band (SB) and uniform band (UB), hence the name “bandsfdp”.

## Installation

You can install the development version of bandsfdp from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("uni-Arya/bandsfdp")
```

## Usage

The standardized and uniform bands require pre-computed Monte Carlo
statistics. These can be downloaded using
`devtools::install_github("uni-Arya/fdpbandsdata")` (approximately
81Mb). The user can also view the code used to generate these tables at
[fdpbandsdata](https://github.com/uni-Arya/fdpbandsdata).

For `tdc_sb()` and `tdc_ub()`, the following inputs are required:

1.  A vector of (non-negative integer valued) rejection `thresholds`.
    Typically only one is used: the rejection threshold of TDC.
2.  A vector of `labels`
    (![-1](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;-1 "-1")
    for a decoy win,
    ![1](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;1 "1")
    for a target win) that are ordered so the corresponding winning
    scores of TDC are decreasing.
3.  A confidence parameter `gamma` (a number between 0 and 1), for a
    `1 - gamma` confidence level. Note that the functions currently
    support `gamma = 0.01, 0.025, 0.5, 0.1, 0.8, 0.5`, but more data can
    be generated using the source code at
    [fdpbandsdata](https://github.com/uni-Arya/fdpbandsdata).
4.  The FDR tolerance `alpha` used in TDC (a number between 0 and 1).

### With a Single Decoy Score

Typically, TDC uses a single decoy score in its competition step. Hence,
both `tdc_sb()` and `tdc_ub()` assume this to be the case by default
(the parameters `c` and `lambda` are both set to `0.5` by default).

Below is an example of how to use these functions. Note that the
`thresholds` are not representative of the actual rejection threshold of
TDC.

``` r
suppressPackageStartupMessages(library(bandsfdp))
set.seed(123)

if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
  thresholds <- c(250, 500, 750, 1000)
  labels <- c(
    rep(1, 250),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
  )
  alpha <- 0.05
  gamma <- 0.05
  
  print(tdc_sb(thresholds, labels, alpha, gamma))
  print(tdc_ub(thresholds, labels, alpha, gamma))
}
#> [1] 0.02000000 0.09453782 0.26825127 0.29575163
#> [1] 0.02400000 0.08823529 0.26315789 0.29084967
```

### With Multiple Decoy Scores

TDC can be extended to use multiple decoys. In that setup, the target
score is competed with multiple decoy scores and the rank of the target
score after competition is used to determine whether the hypothesis is a
target win (label =
$1$),
decoy win
($-1$)
or uncounted
($0$).
The top `c` proportion of ranks are considered winning, the bottom
`1-lambda` losing, and all the rest uncounted. The parameters `c` and
`lambda` must satisfy the following conditions:

1.  $c \leq \lambda$
2.  $c$
    and
    $\lambda$
    are of the form
    $k/(d+1)$
    where
    $d$
    is the number of decoys used and
    $1 \leq k \leq d$
    is an integer.

As an example, if we use
$3$
decoy scores for each hypothesis, we may take
$c$
and
$\lambda$
to be one of
$1/4$,
$1/2$,
or
$3/4$,
subject to
$c \leq \lambda$.
For instance, if
$c = 1/4$,
$H_i$
is labelled as a target win whenever its corresponding target score is
the highest ranked score among all decoys for that hypothesis.

Below is an illustrative example of such a use.

``` r
suppressPackageStartupMessages(library(bandsfdp))
set.seed(123)

if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
  thresholds <- c(250, 500, 750, 1000)
  labels <- c(
    rep(1, 250),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
  )
  alpha <- 0.05
  gamma <- 0.05
  c <- 0.25
  lambda <- 0.25
  
  print(tdc_sb(thresholds, labels, alpha, gamma, c, lambda))
  print(tdc_ub(thresholds, labels, alpha, gamma, c, lambda))
}
#> [1] 0.00800000 0.03991597 0.16298812 0.19444444
#> [1] 0.01200000 0.03781513 0.15449915 0.18627451
```

### Interpolated Bands

All bands are interpolated by default, which requires the computation of
a running maximum. This generally results in a slightly tighter bound,
but at the cost of computational power. We recommend the use of
`interpolate = TRUE`, unless it is too time-consuming.

If one wishes to use non-interpolated bands, the code below shows an
example of such a use.

``` r
suppressPackageStartupMessages(library(bandsfdp))
set.seed(123)

if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
  thresholds <- c(250, 500, 750, 1000)
  labels <- c(
    rep(1, 250),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
  )
  alpha <- 0.05
  gamma <- 0.05
  c <- 0.25
  lambda <- 0.25
  
  print(tdc_sb(thresholds, labels, alpha, gamma, c, lambda, interpolate = FALSE))
  print(tdc_ub(thresholds, labels, alpha, gamma, c, lambda, interpolate = FALSE))
}
#> [1] 0.00800000 0.03991597 1.00000000 1.00000000
#> [1] 0.01200000 0.03781513 1.00000000 1.00000000
```

### Simultaneous FDP bounds

One may also be interested in computing a bound on the FDP of target
wins among the top
$k$
hypotheses for all
$k = 1, \ldots, n$,
where
$n$
is the total number of hypotheses. In this case, the function
`sim_bound()` should be used. This function requires the following
arguments:

- A vector of (ordered) `labels`, confidence parameter `gamma`, and
  competition parameters `c` and `lambda`, as described in the previous
  sections.
- A character argument `type` which is either `"stband"` or `"uniband"`,
  specifying the type of band to be used to compute the simultaneous FDP
  bounds.
- The maximum number of decoy wins considered for the bands `d_max`
  (defaults to `NULL`, in which case it is automatically computed using
  `max_fdp` below).
- The maximal considered FDP for the simultaneous bounds `max_fdp`
  (defaults to `max_fdp = 0.5`).

The arguments `d_max` and `max_fdp` control the rate at which the
simultaneous bounds are increasing. More information is written in the
details section of the R documentation of `sim_bound()`. We also refer
the reader to Section 3 of [Ebadi et
al. (2022)](https://arxiv.org/abs/2302.11837) for more details.

Below is an example of such a use of the function.

``` r
suppressPackageStartupMessages(library(bandsfdp))
set.seed(123)

if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
  set.seed(123)
  labels <- c(
    rep(1, 250),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
  )
  gamma <- 0.05
  sim_bound(labels, gamma, type = "stband")[700:706]
}
#> [1] 0.2402827 0.2416226 0.2416226 0.2416226 0.2416226 0.2416226 0.2429577
```

### Generalized FDP bounds

One may be interested in computing an upper prediction bound on the FDP
among target wins in an arbitrary set
$R$
of hypotheses. In this case, the function `gen_bound()` should be used.
Here, one uses the same arguments as in `sim_bound()`, with an
additional argument `indices` that specifies the set of indices
$R$
for which to compute the upper prediction bound over.

Below is an example of such a use of the function.

``` r
suppressPackageStartupMessages(library(bandsfdp))
set.seed(123)

if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
  set.seed(123)
  labels <- c(
    rep(1, 250),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
    sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
  )
  indices <- c(1:100, 300:400, 600:650)
  gamma <- 0.05
  gen_bound(labels, indices, gamma, type = "stband")
}
#> [1] 0.2546296
```


NEWS.md
# bandsfdp 1.1.0

## New features

- Included function `gen_bound`, which computes an upper prediction bound on the FDP on any arbitrary set of hypotheses $R$.
- Updated README to include an example of use.

## Minor improvements and fixes

- Updated aliases of functions so that `tdc_ub`/`tdc_sb`/`sim_bound` are the primary aliases, instead of `uniband`/`stband`/`simband`. Note both names can still be used to call these functions.
- Fixed a bug with `tdc_ub` and `tdc_sb` that returned a non-trivial bound when the number of decoys was larger than `d_max`

# bandsfdp 1.0.0

-   Initial release of the bandsfdp package.


R/gen_bound.R
#' Generalized band
#'
#' This function computes an upper prediction bound on the FDP among target wins
#' in any set \eqn{R} of hypotheses of TDC. See details for more information.
#'
#' @param labels A vector of (ordered) labels. See details below.
#' @param indices A vector specifying the indices of hypotheses for which an
#' upper prediction bound on the FDP is computed.
#' @param gamma The confidence parameter of the band. Typical values include
#' \code{gamma = 0.05} or \code{gamma = 0.01}.
#' @param type A character string specifying which band to use. Must be one of
#' \code{"stband"} or \code{"uniband"}.
#' @param d_max An optional positive integer specifying the maximum number
#' of decoy wins considered in calculating the bands.
#' @param max_fdp A number specifying the maximum FDP considered by the user in
#' calculating the bands. Used to compute \code{d_max} if \code{d_max} is
#' set to \code{NULL}.
#' @param c Determines the ranks of the target score that are considered
#' winning. Defaults to \code{c = 0.5} for (single-decoy) TDC.
#' @param lambda Determines the ranks of the target score that are
#' considered losing. Defaults to \code{lambda = 0.5} for (single-decoy) TDC.
#'
#' @details In (single-decoy) TDC, each hypothesis is associated to a
#' winning score and a label (1 for a target win, -1 for a decoy win). This
#' function assumes that the hypotheses are ordered in decreasing order of
#' winning scores (with ties broken at random). The argument \code{labels},
#' therefore, must be ordered according to this rule.
#'
#' This function also supports the extension of TDC that uses multiple
#' decoys. In that setup, the target score is competed with multiple decoy
#' scores and the rank of the target score after competition is used to determine whether the
#' hypothesis is a target win (label = 1), decoy win (-1) or uncounted (0).
#' The top \code{c} proportion of ranks are considered winning, the bottom
#' \code{1-lambda} losing, and all the rest uncounted.
#'
#' The threshold of TDC is given by the formula (assuming hypotheses are ordered):
#' \deqn{\max\{k : \frac{D_k + 1}{T_k \vee 1} \cdot \frac{c}{1-\lambda} \leq \alpha\}}{%
#' max{k : (D_k + 1)/max(T_k, 1) \le (1-\lambda)\alpha/c}}
#' where \eqn{T_k} is the number of target wins among the top
#' \eqn{k} hypotheses, and \eqn{D_k} is the number of decoy wins similarly.
#'
#' The argument \code{gamma} sets a confidence level of \code{1-gamma}. Both
#' the uniform and standardized bands require pre-computed Monte Carlo
#' statistics, so only certain values of \code{gamma} are available to use.
#' Commonly used confidence levels, like 0.95 and 0.99, are available.
#' We refer the reader to the README of this package for more details.
#'
#' The argument \code{d_max} controls the rate at which the returned bounds
#' increase: a larger \code{d_max} results in a more conservative bound.
#' If, however, \eqn{D_k + 1} exceeds \code{d_max} for some index \eqn{k}, each target
#' win thereafter is considered a false discovery when computing the bound.
#' Thus it is important that \code{d_max}, chosen a priori, is large enough. Given
#' it is sufficiently large, the precise value of \code{d_max} does not have a
#' significant effect on the resulting bounds (see <https://arxiv.org/abs/2302.11837> for more details).
#'
#' We recommend setting \code{d_max = NULL} so that it is computed automatically
#' using \code{max_fdp}. This argument ensures that \eqn{D_k + 1} never
#' exceeds \code{d_max} when the (non-interpolated) FDP bound on the top
#' \eqn{k} hypotheses is less than \code{max_fdp}.
#'
#' @return An upper prediction bound on the FDP among target wins in the set of
#' hypotheses whose \code{indices} are given as input.
#' @export
#'
#' @examples
#' if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
#'   set.seed(123)
#'   labels <- c(
#'     rep(1, 250),
#'     sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.9, 0.1)),
#'     sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.5, 0.5)),
#'     sample(c(1, -1), size = 250, replace = TRUE, prob = c(0.1, 0.9))
#'   )
#'   indices <- c(1:100, 300:400, 600:650)
#'   gamma <- 0.05
#'   gen_bound(labels, indices, gamma, "stband")
#' }
#'
#' @references Ebadi et al. (2022), Bounding the FDP in competition-based
#' control of the FDR <https://arxiv.org/abs/2302.11837>.
gen_bound <- function(labels, indices, gamma, type,
                      d_max = NULL, max_fdp = 0.5,
                      c = 0.5, lambda = 0.5) {
  if (!type %in% c("stband", "uniband")) {
    stop("Invalid type argument in gen_bound().")
  }
  if (requireNamespace("fdpbandsdata", quietly = TRUE)) {
    n <- length(labels)
    B <- c / (1 - lambda)

    # Name of the table to search for in fdpbandsdata
    name_search <- function(num) {
      format(ceiling(num * 10^5 - 1e-12) / 10^5, nsmall = 5)
    }

    if (type == "stband") {
      # Find the table which contains z (to be used in xi_sb())
      table_name <- paste0(
        "qtable_c",
        name_search(c), "_lam", name_search(lambda)
      )
      table <- get(table_name, envir = getNamespace("fdpbandsdata"))

      xi_sb <- function(d, d_max) {
        if (d + 1 > d_max) {
          return(Inf)
        }
        max(0, floor(z * sqrt(B * (1 + B) * (d + 1)) + B * (d + 1) + 1e-9))
      }
    } else {
      # Find the table which contains u (to be used in xi_ub())
      table_name <- paste0(
        "utable_c",
        name_search(c), "_lam", name_search(lambda)
      )
      table <- get(table_name, envir = getNamespace("fdpbandsdata"))

      xi_ub <- function(d, d_max) {
        if (d + 1 > d_max) {
          return(Inf)
        }
        floor(stats::qnbinom(1 - u, d + 1, prob = (1 / (1 + B))) + 1e-9)
      }
    }

    if (!is.null(d_max)) {
      if (d_max <= 0) {
        stop("Argument d_max is non-positive.")
      }
      if (d_max > 5e4) {
        warning("d_max is greater than 50000. Setting d_max = 50000.")
        d_max <- 5e4
      }

      if (type == "stband") {
        z <- table[d_max, paste0(1 - gamma)]
      } else {
        rho <- table[[1]][d_max, paste0(1 - gamma)]
        sigma <- table[[2]][d_max, paste0(1 - gamma)]
        flip <- table[[3]][d_max, paste0(1 - gamma)]
        u <- ifelse(stats::runif(1) < flip, rho, sigma)
      }
    } else {
      d_max <- 0
      res <- 0
      if (type == "stband") {
        while (res < max_fdp) {
          if (d_max + 1 > n) {
            d_max <- n
            break
          }

          if (d_max + 1 > 5e4) {
            warning("Computed d_max is at least 50000. Setting d_max = 50000.")
            d_max <- 5e4
            break
          }

          z <- table[d_max + 1, paste0(1 - gamma)]

          res <- xi_sb(d_max, d_max + 1) / (n - d_max)
          if (res > max_fdp) {
            break
          }
          d_max <- d_max + 1
        }
      } else {
        while (res < max_fdp) {
          if (d_max + 1 > n) {
            d_max <- n
            break
          }

          if (d_max + 1 > 5e4) {
            warning("Computed d_max is at least 50000. Setting d_max = 50000.")
            d_max <- 5e4
            break
          }

          rho <- table[[1]][d_max + 1, paste0(1 - gamma)]
          sigma <- table[[2]][d_max + 1, paste0(1 - gamma)]
          flip <- table[[3]][d_max + 1, paste0(1 - gamma)]
          u <- ifelse(stats::runif(1) < flip, rho, sigma)

          res <- xi_ub(d_max, d_max + 1) / (n - d_max)
          if (res > max_fdp) {
            break
          }
          d_max <- d_max + 1
        }
      }

      if (d_max == 0) {
        stop("Calculated d_max is 0. The bounds will all be 1.")
      }

      if (type == "stband") {
        z <- table[d_max, paste0(1 - gamma)]
      } else {
        rho <- table[[1]][d_max, paste0(1 - gamma)]
        sigma <- table[[2]][d_max, paste0(1 - gamma)]
        flip <- table[[3]][d_max, paste0(1 - gamma)]
        u <- ifelse(stats::runif(1) < flip, rho, sigma)
      }
    }

    decoys <- cumsum(labels == -1)
    targets <- cumsum(labels == 1)
    xi <- ifelse(type == "stband", xi_sb, xi_ub)

    vbar <- function(index, xi) {
      # If L_i = -1, return \xi_{D_i - 1} or T_i if D_i > d_max
      # Use \xi_{D_i - 1} because our \xi functions map d -> \xi_{d + 1}
      if (labels[index] == -1) {
        return(min(xi(decoys[index] - 1, d_max), targets[index]))
      }
      # Else, return \xi_{D_i} or T_i if D_i + 1 > d_max
      min(xi(decoys[index], d_max), targets[index])
    }

    # Set of target wins among the given indices
    R <- intersect(indices, which(labels == 1))

    # The number of false discoveries in R is at most R
    current_min <- length(R)
    if (current_min == 0) {
      stop("No discoveries among the given set of indices.")
    }

    # Update indices of target wins in top k hypotheses in linear time
    targets_in_top_k <- c()

    # Update the number of false discoveries using interpolation
    for (k in 1:max(R)) {
      if (labels[k] == 1) {
        targets_in_top_k <- c(targets_in_top_k, k)
      }
      possible_min <- vbar(k, xi) + length(setdiff(R, targets_in_top_k))
      if (possible_min < current_min) {
        current_min <- possible_min
      }
    }

    # Return a bound on the number of false discoveries in R
    bound <- current_min / length(R)
    return(bound)

  } else {
    stop(
      "Generalized bands require precomputed data tables. You may choose",
      " to run devtools::install_github(\"uni-Arya/fdpbandsdata\") to install",
      " these tables."
    )
  }
}

#' @rdname gen_bound
genband <- gen_bound
